File: mac-runner/Sources/VoiceRpcProtocol.swift
import Foundation

// MARK: - Voice RPC protocol (task #106, Phase 2 VR companion)
//
// Pure, framework-free wire format for the sidecar<->Swift voice round-trip.
// The Swift half of the protocol; the C# half is mac-runner-host/VoiceRpcProtocol.cs
// and the two must stay in lockstep. Imports only Foundation so it compiles into
// the standalone unit-test binary (same pattern as VoiceTextProcessing).
//
// The Mac sidecar (mac-runner-host) has no native STT/TTS — the macOS Speech /
// AVSpeechSynthesizer frameworks live here in the Swift app (decision #166/#168).
// So /api/voice/query on a Mac host ships audio to us over the existing stdio
// channel: the sidecar writes "voice-*-request" frames on its stdout (which we
// already read), we do the native work, and reply with "voice-*-response" frames
// on its stdin (which its command loop already reads). Each pair is correlated
// by an integer id.

/// A decoded request from the sidecar.
enum VoiceRpcRequest: Equatable {
    /// Base64 WAV (16-bit mono 16 kHz) to transcribe on-device.
    case stt(id: Int, wavBase64: String)
    /// Text to render to a WAV, plus the host's configured voice settings.
    case tts(id: Int, text: String, voiceId: String?, rate: Int, volume: Int)
}

enum VoiceRpcProtocol {
    static let sttRequestPrefix = "voice-stt-request"
    static let ttsRequestPrefix = "voice-tts-request"
    static let sttResponsePrefix = "voice-stt-response"
    static let ttsResponsePrefix = "voice-tts-response"

    static func isVoiceRequest(_ line: String) -> Bool {
        return line.hasPrefix(sttRequestPrefix) || line.hasPrefix(ttsRequestPrefix)
    }

    /// Parse a "voice-*-request <json>" line into a typed request, or nil if the
    /// line isn't a recognizable, well-formed request frame.
    static func parseRequest(_ line: String) -> VoiceRpcRequest? {
        guard let (prefix, json) = splitFrame(line) else { return nil }
        guard let obj = try? JSONSerialization.jsonObject(with: Data(json.utf8)),
              let dict = obj as? [String: Any],
              let id = intValue(dict["id"]) else { return nil }

        switch prefix {
        case sttRequestPrefix:
            guard let wav = dict["wavBase64"] as? String else { return nil }
            return .stt(id: id, wavBase64: wav)
        case ttsRequestPrefix:
            guard let text = dict["text"] as? String else { return nil }
            let voiceId = dict["voiceId"] as? String
            let rate = intValue(dict["rate"]) ?? 0
            let volume = intValue(dict["volume"]) ?? 100
            return .tts(id: id, text: text, voiceId: voiceId, rate: rate, volume: volume)
        default:
            return nil
        }
    }

    /// Build a "voice-stt-response <json>" line. Pass a non-nil `error` to signal
    /// failure (the C# side maps it to a TranscriptionResult.Failure → 500).
    static func sttResponse(id: Int, text: String?, error: String?) -> String {
        return encode(prefix: sttResponsePrefix, body: [
            "id": id,
            "text": text as Any? ?? NSNull(),
            "error": error as Any? ?? NSNull(),
        ])
    }

    /// Build a "voice-tts-response <json>" line. Pass a non-nil `error` to signal
    /// failure (the C# side logs it and omits audio from the response).
    static func ttsResponse(id: Int, wavBase64: String?, error: String?) -> String {
        return encode(prefix: ttsResponsePrefix, body: [
            "id": id,
            "wavBase64": wavBase64 as Any? ?? NSNull(),
            "error": error as Any? ?? NSNull(),
        ])
    }

    // MARK: - WAV writer

    /// Wrap a raw little-endian 16-bit PCM buffer in a canonical 44-byte
    /// RIFF/WAVE header. Mirrors VoiceRpcProtocol.WrapPcm16ToWav on the C# side;
    /// used to package AVSpeechSynthesizer output for the companion to play.
    static func wrapPcm16ToWav(_ pcm: Data, sampleRate: Int, channels: Int) -> Data {
        let bitsPerSample = 16
        let byteRate = sampleRate * channels * bitsPerSample / 8
        let blockAlign = channels * bitsPerSample / 8
        let dataLen = pcm.count

        var header = Data()
        header.append(ascii: "RIFF")
        header.appendLE(UInt32(36 + dataLen))
        header.append(ascii: "WAVE")
        header.append(ascii: "fmt ")
        header.appendLE(UInt32(16))             // fmt chunk size
        header.appendLE(UInt16(1))              // PCM
        header.appendLE(UInt16(channels))
        header.appendLE(UInt32(sampleRate))
        header.appendLE(UInt32(byteRate))
        header.appendLE(UInt16(blockAlign))
        header.appendLE(UInt16(bitsPerSample))
        header.append(ascii: "data")
        header.appendLE(UInt32(dataLen))

        var wav = header
        wav.append(pcm)
        return wav
    }

    // MARK: - Helpers

    private static func splitFrame(_ line: String) -> (prefix: String, json: String)? {
        guard let spaceIdx = line.firstIndex(of: " ") else { return nil }
        let prefix = String(line[line.startIndex..<spaceIdx])
        let json = String(line[line.index(after: spaceIdx)...])
        guard !json.isEmpty else { return nil }
        return (prefix, json)
    }

    private static func encode(prefix: String, body: [String: Any]) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: body, options: [.sortedKeys]),
              let json = String(data: data, encoding: .utf8) else {
            return "\(prefix) {\"id\":0,\"error\":\"encode failed\"}"
        }
        return "\(prefix) \(json)"
    }

    /// JSONSerialization decodes integers as NSNumber; accept either an Int or a
    /// numeric NSNumber so the field survives a round-trip through JSON.
    private static func intValue(_ raw: Any?) -> Int? {
        if let i = raw as? Int { return i }
        if let n = raw as? NSNumber { return n.intValue }
        return nil
    }
}

private extension Data {
    mutating func append(ascii: String) {
        append(contentsOf: ascii.utf8)
    }

    mutating func appendLE(_ value: UInt16) {
        append(UInt8(value & 0xFF))
        append(UInt8((value >> 8) & 0xFF))
    }

    mutating func appendLE(_ value: UInt32) {
        append(UInt8(value & 0xFF))
        append(UInt8((value >> 8) & 0xFF))
        append(UInt8((value >> 16) & 0xFF))
        append(UInt8((value >> 24) & 0xFF))
    }
}

File: mac-runner-host/VoiceRpcProtocol.cs
using System.Buffers.Binary;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace FreeAiSsd.MacRunnerHost;

// Task #106 (Phase 2 VR companion). The Mac sidecar has no native STT/TTS of its
// own — the Swift parent owns the macOS Speech / AVSpeechSynthesizer frameworks
// (decision #166/#168). So /api/voice/query on the Mac host round-trips audio to
// Swift over the existing stdin/stdout channel rather than wiring a .NET engine.
//
// This class is the wire format for that round-trip, kept free of any process /
// pipe / framework dependency so it unit-tests cleanly from tests/ (which already
// references this project). It is the C# half of the protocol; the Swift half is
// mac-runner/Sources/VoiceRpcProtocol.swift and the two must stay in lockstep.
//
// Frames are single newline-delimited lines, "<prefix> <json>", correlated by an
// integer id. Requests flow sidecar -> Swift on the sidecar's stdout (alongside
// the existing "ready:"/"log:" lines); responses flow Swift -> sidecar on the
// sidecar's stdin (alongside the existing "config-update"/"shutdown" commands).

/// <summary>STT request payload: raw audio wrapped as a base64 WAV for the recognizer.</summary>
internal sealed record SttRequestFrame(
    [property: JsonPropertyName("id")] int Id,
    [property: JsonPropertyName("wavBase64")] string WavBase64);

/// <summary>TTS request payload: text plus the host's configured voice settings.</summary>
internal sealed record TtsRequestFrame(
    [property: JsonPropertyName("id")] int Id,
    [property: JsonPropertyName("text")] string Text,
    [property: JsonPropertyName("voiceId")] string? VoiceId,
    [property: JsonPropertyName("rate")] int Rate,
    [property: JsonPropertyName("volume")] int Volume);

/// <summary>STT response payload: the transcript, or a non-null error.</summary>
internal sealed record SttResponseFrame(
    [property: JsonPropertyName("id")] int Id,
    [property: JsonPropertyName("text")] string? Text,
    [property: JsonPropertyName("error")] string? Error);

/// <summary>TTS response payload: a base64 WAV, or a non-null error.</summary>
internal sealed record TtsResponseFrame(
    [property: JsonPropertyName("id")] int Id,
    [property: JsonPropertyName("wavBase64")] string? WavBase64,
    [property: JsonPropertyName("error")] string? Error);

internal static class VoiceRpcProtocol
{
    public const string SttRequestPrefix = "voice-stt-request";
    public const string TtsRequestPrefix = "voice-tts-request";
    public const string SttResponsePrefix = "voice-stt-response";
    public const string TtsResponsePrefix = "voice-tts-response";

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        DefaultIgnoreCondition = JsonIgnoreCondition.Never,
    };

    // ── Request serialization (sidecar -> Swift) ───────────────────────────

    public static string SerializeSttRequest(int id, byte[] wav)
        => $"{SttRequestPrefix} {JsonSerializer.Serialize(new SttRequestFrame(id, Convert.ToBase64String(wav)), JsonOptions)}";

    public static string SerializeTtsRequest(int id, string text, string? voiceId, int rate, int volume)
        => $"{TtsRequestPrefix} {JsonSerializer.Serialize(new TtsRequestFrame(id, text, voiceId, rate, volume), JsonOptions)}";

    // ── Response parsing (Swift -> sidecar) ────────────────────────────────

    public static bool IsVoiceResponse(string line)
        => line.StartsWith(SttResponsePrefix, StringComparison.Ordinal)
        || line.StartsWith(TtsResponsePrefix, StringComparison.Ordinal);

    public static bool TryParseSttResponse(string line, out SttResponseFrame? frame)
        => TryParse(line, SttResponsePrefix, out frame);

    public static bool TryParseTtsResponse(string line, out TtsResponseFrame? frame)
        => TryParse(line, TtsResponsePrefix, out frame);

    private static bool TryParse<T>(string line, string prefix, out T? frame) where T : class
    {
        frame = null;
        if (!line.StartsWith(prefix, StringComparison.Ordinal))
        {
            return false;
        }

        var json = line.Length > prefix.Length ? line[prefix.Length..].TrimStart() : string.Empty;
        if (json.Length == 0)
        {
            return false;
        }

        try
        {
            frame = JsonSerializer.Deserialize<T>(json, JsonOptions);
            return frame is not null;
        }
        catch (JsonException)
        {
            return false;
        }
    }

    // ── PCM -> WAV ─────────────────────────────────────────────────────────

    /// <summary>
    /// Wraps a raw little-endian PCM buffer in a canonical 44-byte RIFF/WAVE
    /// header so the Swift recognizer can load it from a temp file. The endpoint
    /// strips the header off the companion's upload and hands STT the bare PCM
    /// (validated 16-bit mono 16 kHz at RunnerLocalApiService); this restores it.
    /// </summary>
    public static byte[] WrapPcm16ToWav(byte[] pcm, int sampleRate = 16000, short channels = 1, short bitsPerSample = 16)
    {
        int byteRate = sampleRate * channels * bitsPerSample / 8;
        short blockAlign = (short)(channels * bitsPerSample / 8);
        int dataLen = pcm.Length;

        var wav = new byte[44 + dataLen];
        var span = wav.AsSpan();

        Encoding.ASCII.GetBytes("RIFF", span[..4]);
        BinaryPrimitives.WriteInt32LittleEndian(span.Slice(4, 4), 36 + dataLen);
        Encoding.ASCII.GetBytes("WAVE", span.Slice(8, 4));

        Encoding.ASCII.GetBytes("fmt ", span.Slice(12, 4));
        BinaryPrimitives.WriteInt32LittleEndian(span.Slice(16, 4), 16);           // fmt chunk size
        BinaryPrimitives.WriteInt16LittleEndian(span.Slice(20, 2), 1);            // PCM
        BinaryPrimitives.WriteInt16LittleEndian(span.Slice(22, 2), channels);
        BinaryPrimitives.WriteInt32LittleEndian(span.Slice(24, 4), sampleRate);
        BinaryPrimitives.WriteInt32LittleEndian(span.Slice(28, 4), byteRate);
        BinaryPrimitives.WriteInt16LittleEndian(span.Slice(32, 2), blockAlign);
        BinaryPrimitives.WriteInt16LittleEndian(span.Slice(34, 2), bitsPerSample);

        Encoding.ASCII.GetBytes("data", span.Slice(36, 4));
        BinaryPrimitives.WriteInt32LittleEndian(span.Slice(40, 4), dataLen);
        pcm.CopyTo(span.Slice(44));

        return wav;
    }
}
